src/ByteCountingStream.php

```diff
@@ -0,0 +1,63 @@
+<?php
+namespace GuzzleHttp\Psr7;
+
+use Psr\Http\Message\StreamInterface;
+
+/**
+ *  Stream decorator that ensures an expected number of bytes can be read
+ *  from an underlying read stream. \RuntimeException is thrown when
+ *  the underlying stream fails to provide the expected number of bytes.
+ *  Excess bytes will be ignored.
+ */
+class ByteCountingStream implements StreamInterface
+{
+    use StreamDecoratorTrait;
+
+    /** @var int Number of bytes remains to be read */
+    private $remaining;
+
+    /**
+     * @param StreamInterface $stream       Stream to wrap
+     * @param int             $bytesToRead  Number of bytes to read
+     * @throws \InvalidArgumentException
+     */
+    public function __construct(StreamInterface $stream, $bytesToRead)
+    {
+        $this->stream = $stream;
+
+        if (!is_int($bytesToRead) || $bytesToRead < 0) {
+            $msg = "Bytes to read should be non-negative integer, got {$bytesToRead}.";
+            throw new \InvalidArgumentException($msg);
+        }
+
+        if (
+            $this->stream->getSize() !== null &&
+            $bytesToRead > $this->stream->getSize()
+        ) {
+            $msg = "The ByteCountingStream decorator expects to be able to "
+                . "read {$bytesToRead} from a stream, but the stream being decorated "
+                . "only contains {$this->stream->getSize()} bytes.";
+            throw new \InvalidArgumentException($msg);
```

Comment:
>  ... this should be a RuntimeException ...

make sense, `RuntimeException` is more appropriate for this error case, I had a double thought about that as well.

>  ... but this error case seems like something you'd definitely be able to anticipate and recover from ...

I totally agree with the point that we should make it easier for troubleshooting with this recoverable issue, yet an explicit exception for this helps further with a more specific exception name and more flexible message extraction( correct me if I didn't see all the benefits :) ). Refactor error message a bit more would also do the job right? 
Meanwhile, it does make sense to me for having an explicit exception for this.